### DISCUSSION TITLE
Использовать фильтр match для массива исполнителей

### DISCUSSION THREAD

**Discussion by amd64fox** [Chimildic/goofy#55] [Помощь с алгоритмом] [answered]
В одном из скриптов есть такой момент, фильтрую по заданному критерию массив данных:   
```javascript
let tracks = Source.getPlaylistTracks('test', 'id');

Filter.match(tracks, 'test|test2', true);
```
Но еще бы хотел отфильтровать этот массив по артисту т.к. в справке сказано, что в данном случаи `match` фильтрует только по имени и альбому. Но вот как дальше пропустит этот массив через фильтр `match` но уже по исполнителю не понимаю, в справке все прочитал про функцию `getArtists ` перепробовал кучу комбинаций но как применить это к моему массиву или плейлисту так и не въехал...

**Comment by Chimildic** [comment]
В последних версиях исполнитель учитывается

**Старый ответ**
`match` с поднятым флагом инверсии это команда "всё, кроме". Чем здесь поможет просмотр исполнителя?

Сейчас даже не скажу, почему `match` не смотрит на исполнителя. То ли не ко всем сценариям подходит, то ли просто так сложилось.

**Reply by amd64fox** [reply]
> `match` с поднятым флагом инверсии это команда "всё, кроме". Чем здесь поможет просмотр исполнителя?

Этот флаг в данном случаи не чего не значит, это просто пример.

Если конечно это не накладно, то было бы очень удобно, например добавлять или убирать поиск по исполнителю, т.к. действительно это не всегда нужно.

**Comment by Chimildic** [comment, marked as answer]
Добавьте следующую функцию [по инструкции](https://github.com/Chimildic/goofy/discussions/18), чтобы не потерять при обновлении.
```js
Filter.matchAll = function (items, strRegex, invert = false) {
    let regex = new RegExp(strRegex, 'i');
    let filteredTracks = items.filter((item) => {
        if (typeof item == 'undefined') {
            return false;
        } else if (item.hasOwnProperty('album')) {
            return invert ^ (regex.test(item.name.formatName()) || regex.test(item.album.name.formatName()) || regex.test(item.artists[0].name.formatName()));
        }
        return invert ^ regex.test(item.name.formatName());
    });
    Combiner.replace(items, filteredTracks);
};
```

**Reply by amd64fox** [reply]
Cпасибо, только у меня почему то этот код работает как оригинальная match из основной библиотеки, по исполнителю не хочет работать и ошибок нет.  
Я правильно её вызываю `Filter.matchAll` ?

**Reply by amd64fox** [reply]
все,  сам поправил, там скобку ) еще раз спасибо.
вот тут после `.formatName`
```javascript 
regex.test(item.album.name.formatName())
```

**Comment by amd64fox** [comment]
Хотел добавить своё регулярное выражение  в `match`
```javascript 
let regex = new RegExp('(?:[^.\w]|^|^\\W+)' + strRegex + '(?:[^.\w]|\\W(?=\\W+|$)|$)');
```
Но что-то оно не выполняется. Цель была что бы функция match искала только целые слова. 
Если сейчас например в массиве искать ключевое слово  `wind` в результате оно выдаст такое:

![Без имени](https://user-images.githubusercontent.com/62529699/109032524-ca0f2080-76d6-11eb-904a-fb06cc2c5126.jpg)


Или например как контролировать регистр, узнал что за это отвечает флаг `i`, но как  его переключить тут в обратный режим, пробовал его вообще удалить, но поиск все равно регистронезависимый.

**Reply by Chimildic** [reply]
Вам не нужно изменять код `match`. Достаточно написать выражение при вызове. Для наглядности можно использовать [инструмент](https://regex101.com/r/al5nPs/1).
```js
Filter.match(tracks, '(^|\s)wind\s');
```
Про регистр не понимаю, что вы хотите.

**Reply by amd64fox** [reply]
Попробовал подставить выражение как вы сказали, ваше выражение не работает, моё работает но как то полу криво, хотя в [regex101.com ](url) все ищет как надо и с вашим и с моим примером.
 Такой вопрос и как можно эту регулярку поместить в код функции `match`, что бы не было путаницы при вызове ? 

> Про регистр не понимаю, что вы хотите.

Имел ввиду, что сейчас  при поиске регистр может мыть любым `aA` , например если искать `КОТ` то будет найден и `кот` и `Кот`.

**Reply by Chimildic** [reply]
Символ `\s` должен восприниматься как пробел, но почему-то этого не происходит. Заменим на настоящий пробел: `(^| )wind($| )`
Если убрать флаг `i`, регистр будет учитываться

**Reply by amd64fox** [reply]
Ага, с настоящими пробелами получилось, а вот при удалении флага `i` из кода `match`  как и писал выше ситуацию не меняет, ищет также по всему регистру.

**Reply by Chimildic** [reply]
Удаляя `i` меняем и ключевое слово `Wind`.

**Reply by amd64fox** [reply]
Да, естественно менял ключевое слово, но `match` продолжает искать по всему  регистру..

**Reply by Chimildic** [reply]
Далее в коде `match` используется `formatName`. Помимо прочего, понижающий регистр до нижнего. Эта функция используется во многих местах и будет важна при сравнении имен, поиске. Можете модифицировать ее, убрав понижение регистра и добавить в другой файл, появившийся из-за [инструкции](https://github.com/Chimildic/goofy/discussions/18). Также удалив флаг `i`. Затем заменив `formatName` на `formatNameCustom` в этом `match`.
```js

String.prototype.formatNameCustom = function () {
    return this
        .replace(/[',?!@#$%^&*()+-./\\]/g, ' ')
        .replace(/\s{2,}/g, ' ')
        .replace(/ё/g, 'е')
        .trim();
};
```
Зачем вам нужна точность вплоть до регистра?

**Reply by amd64fox** [reply]
Спасибо огромное, регистр завелся.
Дело в том, что последнее время стал активно использовать `mineTracks` и `getCategoryTracks` и если основная часть треков которые приходят, то что я и искал, то так же много всякой шляпы, которая залетает случайно по схожим или похожим словам. И если использовать match в оригинале, что бы отфильтровать все ненужное, то много и хороших треков уходит, а в ручную их не как не удалить т.к. треков много идет, поэтому требовалось более точное сопоставление слов, самое важное это конечно умение искать только целое слово, ну и второстепенное но то же важное это чувствительность к регистру.

**Reply by Chimildic** [reply]
Эти функции ищут по имени плейлиста или категории, на треки не смотрят. Фильтр с регистром может работать когда плейлист и треки с идентичными названиями. Иначе по каким словам можно `match` пускать не совсем понимаю.

**Reply by amd64fox** [reply]
да да, имел ввиду  именно когда уже массив был получен из этих функций.
